sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraService.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.cassandra;
+
+import java.io.Serializable;
+import java.util.List;
+
+import org.apache.beam.sdk.io.BoundedSource;
+
+/**
+ * An interface for real or fake implementations of Cassandra.
+ */
+public interface CassandraService<T> extends Serializable {
+
+  /**
+   * Returns a {@link org.apache.beam.sdk.io.BoundedSource.BoundedReader} that will read from
+   * Cassandra using the spec from
+   * {@link org.apache.beam.sdk.io.cassandra.CassandraIO.CassandraSource}.
+   */
+  BoundedSource.BoundedReader<T> createReader(CassandraIO.CassandraSource<T> source);
+
+  /**
+   * Returns an estimation of the size that could be read.
```

Comment:
There's an asymmetry: createReader takes a CassandraSource, splitIntoBundles takes a CassandraIO.Read, and getEstimatedSizeBytes takes nothing - shouldn't all these 3 functions take the same argument? They're operating on essentially the same information: table and (potentially) query. Also, like I said before, table should not be implicitly passed to CassandraService - it should be passed through function arguments. Otherwise you get even more weird asymmetry: the table stored in CassandraService is redundant with the table passed to splitIntoBundles or createReader. CassandraService itself should encapsulate only the connection to the service itself - not to a particular table.